libyul/optimiser/NodeIdDispenser.h

```diff
@@ -0,0 +1,55 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/ASTNodeRegistry.h>
+
+#include <functional>
+#include <set>
+
+namespace solidity::yul
+{
+
+struct Block;
+class Dialect;
+
+class NodeIdDispenser
+{
+public:
+	using NodeId = ASTNodeRegistry::NodeId;
+	explicit NodeIdDispenser(
+		ASTNodeRegistry const& _labels,
+		std::set<std::string> const& _reserved = {}
+	);
+
+	ASTNodeRegistry const& labels() const { return m_registry.get(); }
+
+	NodeId newId(NodeId parent = 0);
+	NodeId newGhost();
+	ASTNodeRegistry generateNewLabels(Block const& _root, Dialect const& _dialect) const;
+private:
+	NodeId resolveBaseId(NodeId _id) const;
+
+	std::reference_wrapper<ASTNodeRegistry const> m_registry;
+	std::set<std::string, std::less<>> m_reservedLabels;
```

Comment:
I think `std::less` is the default comparison operator, so this could be just `std::set<std::string>`, no?

Reply:
It's a bit convoluted in the code but the point is that I need to be able to use the thing with `string_view`, which only works on the opaque `std::less<>` operator. The default that is set in `std::set<std::string>` is `std::less<std::string>` which can not deal gracefully with `string_view` :)

Reply:
OK, I see. I was not familiar with transparent comparators, but now I understand!